dtm0/recovery.c

```diff
@@ -0,0 +1,2167 @@
+
+	if (!F(recovered)) {
+		/* Wait until the moment where we should start recovery. */
+		do {
+			M0_CO_FUN(CO(fom), heq_await(fom, &F(state), &F(eoq)));
+			if (F(eoq))
+				goto out;
+
+			if (F(state) == M0_NC_ONLINE && !ALL2ALL)
+				break;
+		} while (F(state) != M0_NC_DTM_RECOVERING);
+
+		if (F(state) == M0_NC_ONLINE) {
+			M0_LOG(M0_WARN, "HA told DTM0 service to skip recovery.");
+			F(recovered) = true;
+		}
+	}
+
+	while (!F(recovered)) {
+		M0_CO_FUN(CO(fom), eolq_await(fom, &F(item)));
+		M0_ASSERT(F(item).ei_type != EIT_END);
+
+		recovery_machine_lock(rf->rf_m);
+		remote_rf = recovery_fom_by_svc_find(rf->rf_m,
+						     &F(item).ei_source);
+		if (remote_rf == NULL) {
+			/* XXX: machine is stopping? */
+			recovery_machine_unlock(rf->rf_m);
+			break;
+		}
+		M0_ASSERT(remote_rf != rf);
+		remote_state_update(remote_rf, &F(item));
+
+		F(recovered) = is_local_recovery_completed(rf->rf_m);
+		recovery_machine_unlock(rf->rf_m);
+	}
+
+	M0_BE_QUEUE__FINISH(&rf->rf_eolq, typeof(F(item)));
+
+	/*
+	 * Emit "RECOVERED". It shall cause HA to tell us to transit from
+	 * RECOVERING to ONLINE.
+	 */
+	if (F(recovered)) {
+		recovery_machine_recovered(rf->rf_m,
+					   &rf->rf_tgt_proc, &rf->rf_tgt_svc);
+
+		do {
+			M0_CO_FUN(CO(fom), heq_await(fom, &F(state),
+						     &F(eoq)));
+			M0_ASSERT(ergo(!F(eoq), F(state) == M0_NC_ONLINE));
+		} while (!F(eoq));
+	}
+
+out:
+	M0_BE_QUEUE__FINISH(&rf->rf_eolq, typeof(F(item)));
+	m0_fom_phase_set(fom, RFS_DONE);
+}
+
+static void recovery_fom_coro(struct m0_fom *fom)
+{
+	struct recovery_fom *rf = M0_AMB(rf, fom, rf_base);
+
+	M0_CO_REENTER(CO(fom));
+
+	addb2_relate(&rf->rf_m->rm_sm, &fom->fo_sm_phase);
+
+	if (rf->rf_is_local)
+		M0_CO_FUN(CO(fom), local_recovery_fom_coro(fom));
+	else
+		M0_CO_FUN(CO(fom), remote_recovery_fom_coro(fom));
+}
+
+static int recovery_fom_tick(struct m0_fom *fom)
+{
+	int rc;
+	M0_CO_START(CO(fom));
+	recovery_fom_coro(fom);
+	rc = M0_CO_END(CO(fom));
+	M0_POST(M0_IN(rc, (0, M0_FSO_AGAIN, M0_FSO_WAIT)));
+	return rc ?: M0_FSO_WAIT;
+}
+
+#undef F
+
+M0_INTERNAL void
+m0_ut_remach_heq_post(struct m0_dtm0_recovery_machine *m,
+		      const struct m0_fid             *tgt_svc,
+		      enum m0_ha_obj_state             state)
+{
+	struct recovery_fom *rf = recovery_fom_by_svc_find_lock(m, tgt_svc);
+	M0_ASSERT_INFO(rf != NULL,
+		       "Trying to post HA event to a wrong service?");
+	heq_post(rf, state);
+}
+
+M0_INTERNAL void
+m0_ut_remach_populate(struct m0_dtm0_recovery_machine *m,
+		      struct m0_conf_process          *procs,
+		      const struct m0_fid             *svcs,
+		      bool                            *is_volatile,
+		      uint64_t                         objs_nr)
+{
+	uint64_t i;
+
+	for (i = 0; i < objs_nr; ++i)
+		M0_ASSERT(recovery_fom_add(m, procs + i,
+					   svcs + i, is_volatile[i]) == 0);
+}
+
+M0_INTERNAL void
+m0_dtm0_recovery_machine_redo_post(struct m0_dtm0_recovery_machine *m,
```

Comment:
Please document this function here or in the header.

Reply:
2:Working on it.